packages/flutter_tools/lib/src/asset.dart

```diff
@@ -1030,25 +1031,32 @@ class _AssetDirectoryCache {
     if (_cache.containsKey(assetPath)) {
       return _cache[assetPath]!;
     }
-
-    final List<FileSystemEntity> entitiesInDirectory = _fileSystem.directory(directory).listSync();
-
+    if (!_variantsPerFolder.containsKey(directory)) {
+      _variantsPerFolder[directory] = _fileSystem.directory(directory)
+        .listSync()
+        .whereType<Directory>()
+        .where((Directory dir) => _assetVariantDirectoryRegExp.hasMatch(_approximateBasename(dir, _fileSystem)))
+        .expand((Directory dir) => dir.listSync())
+        .whereType<File>()
+        .toList();
+    }
     final File assetFile = _fileSystem.file(assetPath);
-    final List<String> pathsOfVariants = <String>[
+    final List<File> potentialVariants = _variantsPerFolder[directory]!;
+    final String basename = _approximateBasename(assetFile, _fileSystem);
+    return _cache[assetPath] = <String>[
       // It's possible that the user specifies only explicit variants (e.g. .../1x/asset.png),
       // so there does not necessarily need to be a file at the given path.
       if (assetFile.existsSync())
         assetPath,
-      ...entitiesInDirectory
-        .whereType<Directory>()
-        .where((Directory dir) => _assetVariantDirectoryRegExp.hasMatch(dir.basename))
-        .expand((Directory dir) => dir.listSync())
-        .whereType<File>()
-        .where((File file) => file.basename == assetFile.basename)
+      ...potentialVariants
+        .where((File file) => _approximateBasename(file, _fileSystem) == basename)
         .map((File file) => file.path),
     ];
-
-    _cache[assetPath] = pathsOfVariants;
-    return pathsOfVariants;
   }
 }
+
+// .basename is surprisingly slow as it hits the real file system to ask for the
+// current directory.
+String _approximateBasename(FileSystemEntity entity, FileSystem fileSystem) {
```

Comment:
This was a large part of the performance fix. Even though we cache the Path context on our error handling filesystem, at this point entity.fileSystem returns localFileSystem instead of ErrorHandlingFIleSystem.